Avoid duplicating search results

weiquu opened the pull request:
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [x] Feature addition or enhancement
- [ ] Code maintenance
- [ ] DevOps
- [ ] Improve developer experience
- [ ] Others, please explain:

<!--
  If this pull request is addressing an issue, link to the issue: "Fixes #xxx" or "Resolves #xxx"

  If the pull request completely addresses the issue, use one of the issue closing keywords.
  https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

  Otherwise, elaborate further on the rationale of this pull request as needed
-->

Resolves #311 
Resolves #1972 

**Overview of changes:**
Add in a check to ensure that headers at the very top of the page which are an exact match of the title are not displayed in the search results. Also changed the design of the results dropdown to avoid duplicating results and to make clear which results are page titles and which are headers within the page.

**Anything you'd like to highlight/discuss:**
Screenshots are in the comments below!

**Testing instructions:**
For the main test of the implementation, create a page with the page title the same as the first and second heading. The first heading should not show up, while the second heading does. 

For testing of heading keywords, create a page with the page title the same as the first heading, and with the first heading having some keyword not contained in heading/page title. Upon searching the keyword, only the page title should turn up (i.e. without the first heading below it).

For testing of UI/UX, use long headings and keywords, with some having no spaces within the words as well.

For page keywords and heading indexing, please see [this comment](https://github.com/MarkBind/markbind/pull/1968#discussion_r939712691) for the issues.

<!--
  Any special testing instructions **not including** our automated tests.
-->

**Proposed commit message: (wrap lines at 72 characters)**
Update displaying of search results

Search results currently show a duplicate column for headings.
Topmost headings are also included in the search results even if
they are exactly the same as the title.

This duplication is redundant and could appear to the user as
being buggy due to the repetition.

To avoid these issue, let's
* remove the duplicate results column for headings
* remove the duplication of search results if the topmost heading
matches the title exactly

<!--
  See this link for more info on how to write a good commit message:
  https://oss-generic.github.io/process/docs/FormatsAndConventions.html#commit-message

  As best as possible, write a succinct commit title in 50 characters

|---------This is the width of 50 chars----------|
|-----------This is the width of 72 chars for your reference-----------|
-->

---

**Checklist:** :ballot_box_with_check:

<!-- Leave non-applicable items unchecked -->

- [ ] Updated the documentation for feature additions and enhancements
- [ ] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No unrelated changes <!-- Its tempting, but increases the reviewer's work, and really pollutes the commit history =( -->

<!--
  We'll try our best to get to your PR within a week.
  If we haven't gotten to it then, or if your pull request resolves an urgent item, feel free to give us a ping!
-->

ong6 commented on packages/vue-components/src/Searchbar.vue:
```diff
@@ -166,10 +166,120 @@ export default {
         }
       });
 
+      // using distance with adjacent transpositions
```
Would be nice if you could explain what exactly this function does, since not everyone may know this algo.

ong6 reviewed (commented):
Looks pretty good, nice work here! Just clean up the comments and it'll LGTM.

weiquu commented:
Thanks @ong6! Have removed the commented out code and added an explanation.

ong6 reviewed (approved):
LGTM!

jonahtanjz reviewed (commented):
Nice work @weiquu!

You might want to add a note in the user guide (under [Making the Site Searchable](https://markbind-master.netlify.app/userguide/makingthesitesearchable#making-the-site-searchable) section) so that users are aware of this behaviour :)

tlylt commented on packages/vue-components/src/Searchbar.vue:
```diff
@@ -166,10 +166,86 @@ export default {
+        }
+
+        let cost = 0;
+        let db = 0;
```
Might be a dumb question: is `db` supposed to mean something?

tlylt commented on packages/vue-components/src/Searchbar.vue:
```diff
@@ -166,10 +166,86 @@ export default {
+
+        // For anything else, we check if the edit distance is less than or equals to the limit set above.
+        const dist = calcDamerauLevenshtein(heading, title);
+        if (dist <= limit) {
+          return true;
+        }
+
+        return false;
```
```suggestion
        return dist <= limit;
```

tlylt commented on packages/vue-components/src/Searchbar.vue:
```diff
@@ -166,10 +166,86 @@ export default {
+          return false;
+        }
+
+        const heading = page.headings[0].heading.text.toLowerCase().trim();
```
I think we can be more defensive here and check for the existence (right at the beginning) of headings and their length before proceeding

tlylt commented on packages/vue-components/src/Searchbar.vue:
```diff
@@ -166,10 +166,86 @@ export default {
+        return matrix[s1.length + 1][s2.length + 1];
+      }
+
+      function isDuplicateResult(page) {
```
Perhaps we can give this a better name? Or minimally a JSdoc, because reading it now "isDuplicateResult(page)" is a little vague, what does "result" mean here?
What do you think?

Perhaps what I can suggest is: hasSimilarPageTitleAndHeading(page)

ang-zeyu commented:
Thanks for the PR @weiquu. Is this what you had in mind? (see OP description) @damithc 

I'm not sure if these checks (in particular edit distance) would be generally applicable enough, some examples:

multilingual:

```
title: any 3 chinese characters

# any other 3 chinese characters
```

even with latin / ascii text, there are issues:

```
title: Today

# Todo // semantics are very different, both should be presented
```

We can scale the edit distance limit by the string length, but that is still susceptible:

```
// if we adopted limit <= 1 for strings of length <= 5,
title: Today

# Koday // don't really have the time now to think of a proper example, but imagine this is a product name, etc. =|
```


I'm supposing this issue was raised because in many sites, the *level 1 heading* in particular is often a "repeat" of the title, so I wonder if we should instead be supporting a range for the [heading indexing level](https://markbind.org/userGuide/siteJsonFile.html#headingindexinglevel).

e.g. `"headingIndexingLevel": "2-4"`,


exceptions can be dealt with by using the `always/no-index` class we already [have](https://markbind.org/userGuide/makingTheSiteSearchable.html#including-or-excluding-headings)

tlylt commented:
> Thanks for the PR @weiquu. Is this what you had in mind? (see OP description) @damithc
> 
> I'm not sure if these checks (in particular edit distance) would be generally applicable enough, some examples:
> 
> multilingual:
> 
> ```
> title: any 3 chinese characters
> 
> # any other 3 chinese characters
> ```
> 
> even with latin / ascii text, there are issues:
> 
> ```
> title: Today
> 
> # Todo // semantics are very different, both should be presented
> ```
> 
> We can scale the edit distance limit by the string length, but that is still susceptible:
> 
> ```
> // if we adopted limit <= 1 for strings of length <= 5,
> title: Today
> 
> # Koday // don't really have the time now to think of a proper example, but imagine this is a product name, etc. =|
> ```
> 
> I'm supposing this issue was raised because in many sites, the _level 1 heading_ in particular is often a "repeat" of the title, so I wonder if we should instead be supporting a range for the [heading indexing level](https://markbind.org/userGuide/siteJsonFile.html#headingindexinglevel).
> 
> e.g. `"headingIndexingLevel": "2-4"`,
> 
> exceptions can be dealt with by using the `always/no-index` class we already [have](https://markbind.org/userGuide/makingTheSiteSearchable.html#including-or-excluding-headings)

I agree with the point raised..have similar concerns about the effectiveness of the edit distance check. On a separate note, seems like the current implementation won’t be able to support languages other than English (iirc)

damithc commented:
> I'm supposing this issue was raised because in many sites, the _level 1 heading_ in particular is often a "repeat" of the title.

Yes, that's the specific problem I'm trying to fix. 

Here is an example:
![image](https://user-images.githubusercontent.com/1673303/180275072-c94b14a6-d539-46b3-bec4-0920ff4f15a5.png)

code:
```html
<frontmatter>
  title: Binary Search Tree
</frontmatter>

# Binary Search Tree

> More content to be added
```

I think the problem case occurs when the page title is **repeated exactly** as the very first heading, at the **very top of the page** (as the case above). In such a case, the two search results goes to the exact same place in the site. Hence, the second result is redundant and an eyesore too (at first glance it might even look like search results are buggy as the same words are repeated thrice in the same place).

damithc commented:
On a separate note, these two columns of the search results almost always show the exact same content? Is that intended?
![image](https://user-images.githubusercontent.com/1673303/180277500-e394a1d1-8e9a-4ea5-9880-463d632f0c7f.png)

weiquu commented:
Thanks everyone for the feedback! Would a better solution be to simply check if the level 1 heading is contained within the page title and remove it if so (regardless of whether multiple headings are found by the search)? Then we can do away with the edit distance, which removes the language issues raised by @ang-zeyu and also solves the issue that @damithc wants to fix.

Not too sure about the separate issue as the dropdown looked like that previously! I can try making it into a single column if needed (:

damithc commented:
> Not too sure about the separate issue as the dropdown looked like that previously! I can try making it into a single column if needed (:

If this is an issue, it is likely a separate issue, in which case we should create a separate issue and fix it separately.



> Would a better solution be to simply check if the level 1 heading is contained within the page title and remove it if so

_contained within_ or _same as_?
Also, the problem occurs only when the offending heading is at the top of the page. If it is in a different location, the two search results point to two different places in the page, and hence, not technically duplicates.

ang-zeyu commented:
> which removes the language issues raised by @ang-zeyu

not quite unfortunately, a substring check is also problematic. In general I don't think we can avoid making assumptions (which we should avoid) about the page content with the current solutions in the PR:

```
// e.g.
title: 3D Procedural Game Engine Using OpenGL

# OpenGL
```

weiquu commented:
Oki I'll change it to checking if the heading is at the top of the page and is exactly the same as the title!

ang-zeyu commented:
> I'll change it to checking if the heading is at the top of the page

Hmm that's certainly one way, but detecting if the heading is at the top may be slightly difficult. (html comments, frontmatter), and/or not very performant (extracting the html text then doing something like a `startsWith` check)

> so I wonder if we should instead be supporting a range for the [heading indexing level](https://markbind.org/userGuide/siteJsonFile.html#headingindexinglevel).

What do you think about this?

ping @damithc @MarkBind/active-devs as well

damithc commented:
> > so I wonder if we should instead be supporting a range for the [heading indexing level](https://markbind.org/userGuide/siteJsonFile.html#headingindexinglevel).
> 
> What do you think about this?

Because the problem is specific to some pages only, omitting all H1 is not an option. We can let the author configure it at page level but that is a lot of work for the author. Plus the author might not be aware of this problem at all, or which pages are affected. So, a 'it just works' solution that the author doesn't even have to know about is best but I realize performance can be affected.

As @ang-zeyu noted, it's good if the solution works for Algolia as well, but seems like Algolia search results are not affected by this issue. Apparently, the page title is not shown in Algolia search results. Should we do something similar?

![image](https://user-images.githubusercontent.com/1673303/180592633-0b469247-bffa-4847-9d71-1825be799eef.png)

![image](https://user-images.githubusercontent.com/1673303/180592642-33c6cbbf-adf8-4fe7-8d46-205b697a1a90.png)

Also note that #1972 makes this problem appear worse (i.e., repeats the same thing thrice). Perhaps we tackle that one first?

ang-zeyu commented:
> > I'll change it to checking if the heading is at the top of the page
>
> Hmm that's certainly one way, but detecting if the heading is at the top may be slightly difficult. (html comments, frontmatter), and/or not very performant (extracting the html text then doing something like a `startsWith` check).
>
> *just a rough estimate though, if there's an easy way to do it that would be great.

@weiquu To be clear again I might _very, very_ well be wrong here, feel free to dig around the html processing [source](https://github.com/MarkBind/markbind/blob/master/packages/core/src/html/NodeProcessor.js).

> but seems like Algolia search results are not affected by this issue.

That might be another solution. I think Algolia always uses `h1` as its top level item, iirc Google does this to some degree as well.

But I think this comes after the first solution (detecting if heading is at the top of the page). We can also prioritise top-level, first `h1` but fallback to `title`.

---

On a slightly unrelated note... I liked Algolia docsearch v2's UI better. Having level 2 / 3 items (anchors) displayed on the same level as level 0 items (h1) is visually confusing for me and rather unconventional to understand the SERP hierarchy. The hashtag icon does not help. =|
*To be clear, this random rant is separate from the idea of using `h1` as a level 0 item.

weiquu commented:
Fixed the issue in #1972 by expanding out the left column and adding a hashtag to indicate that the result is a heading within the page. Also added a quick check to ensure that headings at the very top of the page which match the title exactly do not get shown in the search results. Below is an example with the page title being `Using Plugins` and 4 headings (in order) `Using Plugins`, `Using Plug-ins`, `Using Plugins`, and `Using Plug-in`. As seen, the first heading is removed from the search results:
![Search Results](https://user-images.githubusercontent.com/48304907/182022098-0838e352-1d71-4f95-8d52-6d84ed44d758.png)

Here is the search results with long headings:
![Long Heading](https://user-images.githubusercontent.com/48304907/182022113-98ee57b1-4b1a-4f4e-ac2e-724f0ab0f7d3.png)

ang-zeyu commented:
> Also added a quick check to ensure that headings at the very top of the page which match the title exactly do not get shown in the search results. Below is an example with the page title being Using Plugins and 4 headings (in order) Using Plugins, Using Plug-ins, Using Plugins, and Using Plug-in. As seen, the first heading is removed from the search results:

I was thinking of a few other edge cases we can deal with later on by looking at the page content at build time instead, although the implementation might be a little more complex.


e.g.
```
lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem 

... very long content ...

# Ipsum
```

```
# Lorem {.no-index}

...

# Ipsum
```

Unless there's any objections, I think this simple check is a good enough stopgap for most cases, and since document recall (which is more critical) ultimately isn't affected by the exact check 🙂👍 

> Fixed the issue in https://github.com/MarkBind/markbind/issues/1972 by expanding out the left column and adding a hashtag to indicate that the result is a heading within the page. 

Thanks for the fix. Let's move this to another PR though, the scope of the change is very different.

You may also need to check if this plays correctly and nicely with [keywords](https://markbind.org/userGuide/makingTheSiteSearchable.html#keywords).

ang-zeyu commented on packages/vue-components/src/Searchbar.vue:
```diff
@@ -134,6 +134,7 @@ export default {
         if (totalPageMatches > 0) {
           const pageHeadings = [];
+          let first = true;
           Object.entries(headings).forEach(([id, text]) => {
```
would `idx === 0` (from `forEach`) work too?

weiquu commented:
Okay have updated the code to use index instead! I've also tested it with keywords (page and headings), and based on my tests it works fine.

tlylt commented:
> Okay have updated the code to use index instead! I've also tested it with keywords (page and headings), and based on my tests it works fine. Have also updated the commit message (:

Hi @weiquu, a quick note on the commit message in case you missed it:
![image](https://user-images.githubusercontent.com/41845017/182563315-8b170f9e-1edf-4b91-b5e3-37f600a36ec1.png)

Feel free to read up on https://oss-generic.github.io/process/docs/FormatsAndConventions.html#commit-message for our commit message format. I think this change is non-trivial, perhaps we can update the commit title and body to summarize the change done 🦾

damithc commented:
> Feel free to read up on https://oss-generic.github.io/process/docs/FormatsAndConventions.html#commit-message for our commit message format.

Moved to https://se-education.org/guides/conventions/git.html (OSS-generics is now defunct).

weiquu commented:
Thanks, have updated the commit message accordingly!

ang-zeyu commented on packages/vue-components/src/SearchbarPageItem.vue:
```diff
@@ -1,8 +1,10 @@
-      <span v-html="item.heading.text"></span>
+      <span v-html="highlight(item.heading.text, value)"></span>
     </div>
+    <!--
```
> You may also need to check if this plays correctly and nicely with [keywords](https://markbind.org/userGuide/makingTheSiteSearchable.html#keywords).
> > I've also tested it with keywords (page and headings), and based on my tests it works fine.

I actually meant moving the keywords to a separate location (e.g. below the heading as smaller, muted text), not removing them entirely.

But this looks ok to me as well (as long as it turns up).

Let's remove the dead code (one more below as well)

On a related note https://github.com/MarkBind/markbind/issues/205 would change the design completely again... but don't think I'm going to be able to complete that anytime soon.

weiquu replied:
Ah sorry for misunderstanding! I've added in the keywords to the search results, looks like this:

![keywords long](https://user-images.githubusercontent.com/48304907/183283270-c49716c5-eb02-4c82-861c-20c04337dc5c.png)

Have also removed the dead code (:

tlylt replied:
Hi @weiquu, thanks for the update. A gentle reminder to provide testing instructions in the future, so that reviewers can get back to you quicker:) For the screenshot you posted above, I tried it on the PR preview in https://deploy-preview-1968--markbind-master.netlify.app/ and then realized that it won't work because you have added some keywords locally to show this behavior.

Here's some screenshots of existing keywords and how they look like with this update:
- hide minimized panels
![image](https://user-images.githubusercontent.com/41845017/183283705-40038efe-f1ff-4319-b472-5666a2b9b645.png)
- regress
![image](https://user-images.githubusercontent.com/41845017/183283717-e8d459fa-bef9-4635-8645-d2401e1c8774.png)

Are the words suppose to start on a newline?

Lastly, may I double check if the content here is still valid after your change @weiquu ?
https://markbind.org/userGuide/makingTheSiteSearchable.html
Specifically:
- does keyword still work for heading and an entire page?
- does the `.always-index or .no-index` still work,  in particular, if the heading is the very first one?

weiquu replied:
Thanks for the reminder! Will update the testing instructions.

For the newline issue, can I check if you went into the PR preview and directly typed in the search words? I tried and there's no newline for me (e.g. regress below).
![regress](https://user-images.githubusercontent.com/48304907/183307482-c960055f-a9e8-46d4-a616-330dc7a64321.png)

I've checked through the content in the page linked, responding to the 2 pointed out:

For the `.always-index` and `.no-index`, I've tested with the code below:
```
<span id="title" class="d-none">Using Plugins</span>

<frontmatter>
  title: "Using Plugins"
</frontmatter>

###### Using Plugins {.always-index}

# Using Plugins {.no-index}
```
With this code, both don't show up (as expected), but if the positions of the 2 headings are swapped around (h1 then h6), then both don't show up again since the h6 heading is taken as the first indexed heading.

For the keywords, they do also work with the heading. However, I'm not too sure if I'm doing something wrong with the  page heading since it doesn't seem to work even if I revert back to the main build. With this code, I'm not seeing anything coming up with search word `plugin` even though I believe the expected result should be seeing the page title `Formatting Contents`:
```
<span id="title" class="d-none">Formatting Contents</span>

<frontmatter>
  title: "Formatting Contents"
  keywords: plugin
</frontmatter>

# Formatting Contents (Advanced)
```

Would appreciate any advice on the above 2 issues!

ang-zeyu replied:
> does the .always-index or .no-index still work, in particular, if the heading is the very first one?
>
> > With this code, both don't show up (as expected), but if the positions of the 2 headings are swapped around (h1 then h6), then both don't show up again since the h6 heading is taken as the first indexed heading.

I might be lacking some context here but imo this change should have nothing to do with these attributes. (current behaviour is fine)

E.g.

```
<frontmatter>
  title: "Using Plugins"
</frontmatter>

// Should not show. What we index =/= user friendliness of the SERP we generate.
# Using Plugins {.always-index}
```

> does keyword still work for heading and an entire page?
>
> > For the keywords, they do also work with the heading. However, I'm not too sure if I'm doing something wrong with the page heading since it doesn't seem to work even if I revert back to the main build. With this code, I'm not seeing anything coming up with search word plugin even though I believe the expected result should be seeing the page title Formatting Contents:

That's a good catch. I think we can add to the check something along the lines of `!(idx === 0 && ... && !keywords.length)`.

ang-zeyu replied:
On a separate note, a 🔑 icon conveys to me *security* - e.g. a password, authentication, secret key, etc. Might be confusing to readers what exactly the words following are.

I think we ought to swap up or remove the icon. 🤔 

If there's another suitable icon, let's limit it to displaying just once too so it dosen't get too repititive.

tlylt replied:
> For the newline issue, can I check if you went into the PR preview and directly typed in the search words? I tried and there's no newline for me (e.g. regress below).

@weiquu see gif below, seems like it depends on window size:
![OQyxxAH6uv](https://user-images.githubusercontent.com/41845017/183543852-a00134c4-349d-4668-a4b2-85b02d6615b6.gif)

> On a separate note, a 🔑 icon conveys to me security

How about a 🎯 ?

> I might be lacking some context here but imo this change should have nothing to do with these attributes.

Was checking if these attributes will be affected by this implementation, because of how we  ignore duplicates (what if the duplicates have been marked with the always-index, no-index, does it still work correctly)

> For the keywords, they do also work with the heading. However, I'm not too sure if I'm doing something wrong with the page heading since it doesn't seem to work even if I revert back to the main build. 

Yes seems like it is an existing issue.

ang-zeyu replied:
> 🎯

This conveys target - which means a goal, okr to me personally.

Would a muted search icon work?

On further thought I'm favouring removing the icon more strongly as well, rationale is that the author can also add unicode emojis themselves iirc.

ang-zeyu replied:
> Was checking if these attributes will be affected by this implementation, because of how we ignore duplicates (what if the duplicates have been marked with the always-index, no-index, does it still work correctly)

🙆‍♂️

weiquu commented:
> I think we can add to the check something along the lines of `!(idx === 0 && ... && !keywords.length)`

Have added this in!

> Yes seems like it is an existing issue.

Okay, I'll leave it as is for this PR, should I open a new issue for it?

> On further thought I'm favouring removing the icon more strongly as well, rationale is that the author can also add unicode emojis themselves iirc.

Have removed the icon as well, do let me know if it works (or if I should try making the text smaller)

tlylt commented:
@weiquu 

> Okay, I'll leave it as is for this PR, should I open a new issue for it?

Sure, sounds good to me.

On a separate note, though there's no longer the icon issue (because there's no more icon ,-,), the `#` symbol also faces the same problem. Is this display of symbol and text on different line the expected behavior?
![image](https://user-images.githubusercontent.com/41845017/186635530-947b3b18-0125-4698-a7c9-46c5aa6124f7.png)

weiquu commented:
Okay have submitted the issue! I think I've also fixed the line break issue (wrong calculation of width)

tlylt commented on packages/vue-components/src/SearchbarPageItem.vue:
```diff
@@ -1,15 +1,13 @@
       <small v-for="(keyword, index) in item.keywords" :key="index">
-        <span v-html="highlight(keyword, value)"></span>
-        <br />
+        <div class="heading-text">
```
I see you have this div and class "heading-text" to ensure that the keyword starts on a new line. Wonder if you have tried simplifying it to `<div v-html="highlight(keyword, value)"></div>` and do without the class? Seems like it works for most cases. What do you think @weiquu?

weiquu replied:
Thanks I think it works well! Have changed it

tlylt replied:
Hi @weiquu, sorry for going back and forth on this one (the rest of the PR LGTM). On second thought I feel that this update might not be very semantic because small is an inline element while div is a block element, so inline should technically not contain a block.

 I am suggesting to change to the following to avoid possible complication down the line:

```
    <div class="heading-text">
      <span v-html="highlight(item.heading.text, value)"></span>
      <div
        v-for="(keyword, index) in item.keywords"
        :key="index"
        class="keyword-text"
        v-html="highlight(keyword, value)"
      ></div>
    </div>

.keyword-text {
  font-size: smaller;
}
```

(P.S then again, the rule is broken in the dropdown level outside of this, where an inline `<a>` wraps the `<div>`, but things still work fine...) pinging @ang-zeyu @jonahtanjz for advice on this 👁️  
![image](https://user-images.githubusercontent.com/41845017/188251068-e5401cf8-0f9b-4cc8-a2d5-bd0dd6c2b71f.png)

ang-zeyu replied:
> (P.S then again, the rule is broken in the dropdown level outside of this, where an inline <a> wraps the <div>, but things still work fine...) pinging @ang-zeyu @jonahtanjz for advice on this 👁️

This is actually ok (*not* the `<small><div>...` though).

Some links: 
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#properties 
https://html.spec.whatwg.org/#transparent-content-models
 to validate the markup - https://validator.w3.org/nu/#textarea 
You can also check out how algolia docsearch structures its html - https://docusaurus.io/ (just using docusaurus as a convenient example)

I am nowhere an expert in this area too 😅😂  but in general also note browsers don't comply with the spec completely; Failing gracefully is as is the case here is often the better option for the end user. https://stackoverflow.com/questions/64821044/how-browser-restructure-the-invalid-html-structure-is-there-any-rules-or-approa

ang-zeyu replied:
SSR hydration assertions also do not check everything, likely just some of the most common ones like `<table>`s that browsers autocorrect aggressively

weiquu replied:
Thanks for the links! Have edited according to @tlylt's suggestion

tlylt reviewed (approved):
Thanks @weiquu, LGTM
